Added a save/load feature to Result

dcmckayibm opened the pull request:
Added functionality to the Result class to save/load the qobj and result to a JSON file. 

## Description
The save creates a new dictionary that combines qobj,result and a user-defined metadata dictionary. This is then saved to a JSON file. There is a function which can create a time-stamped filename. Because the 'local_unitary_simulator' backend creates a result dictionary with complex numpy arrays I added some hidden functions to the Result class to convert those into lists. 

## Motivation and Context
Currently there is no way to save results. This is needed functionality for running experiments where we need to keep a detailed record of past results and for analyzing results at a later date.

## How Has This Been Tested?
I saved and loaded Results from the different backends.  

Code will look something like this (where qpresult is the result output from an execution)
`qpresult.save(qpresult.save_datestr('','test.json'),x_vals)`
`qpresult2 = Result()`
`qpresult2.load('2017_10_11_00_52_test_1.json')`

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

dcmckayibm commented:
I added some unit tests.

diego-plan9 commented on qiskit/_result.py:
```diff
@@ -253,4 +318,88 @@ def get_qubitpol_vs_xval(self, xvals_dict=None):
+        self._convert_qobj_to_json(master_dict['result'])
+
+        #if the filename has .json appended strip it off
+        if filename[-5:].lower()=='.json':
```
In the spirit of the comment for `save_datestr()`, would it be possible to assume that the user is responsible for determining the filename and avoid trying to extract the `.json` part?

diego-plan9 commented on qiskit/_result.py:
```diff
@@ -253,4 +318,88 @@ def get_qubitpol_vs_xval(self, xvals_dict=None):
+
+        return filename+append_str+'.json'
+
+    def load(self, filename):
```
Related to the comment on the `Result.__init__()` changes, can we change this method to a class method that returns the newly constructed `Result`, renaming it a bit in the process? Something along the lines of:

```python
@classmethod
def from_json(cls, filename):
    ...
    return cls(result, qobj)
```

So the usage would be:
```python
my_loaded_result = Result.from_json('my_json_file.json')
```

The main drawback would be that `metadata` would not be returned, but the proposed change would make sure that all `Result` instantations would go through the constructor, which reduces the complexity by streamlining the instantiation into a single point which would be easy to change in the future and avoids having `Results` on an invalid state (ie. with `None` attributes). Even if that state would be only for a short while on its lifetime, doing it through the constructor avoids having to treat the loading as a separate case and reduces the chance of forgetting to update it in the future and similar implications.

For solving the issue of retrieving `metadata`, I'm unsure about the best option, but perhaps a reasonable option would be to have a `get_metadata_from_json()` method (ideally outside the class, as a function module, as it does not really use `Result`) that only parses the file and returns the converted `metadata` field.

diego-plan9 commented on qiskit/_result.py:
```diff
@@ -253,4 +318,88 @@ def get_qubitpol_vs_xval(self, xvals_dict=None):
+        Args:
+            filename: save path
+            metadata (optional): Add another dictionary with custom data for the result (eg fit results)
+
```
Can we please also add better-worded version of:
```
Note:
    The resulting files are version-specific, and it is not guarantees that a .json file saved with
    a specific version of the SDK will be cleanly loaded in subsequent versions ...
```

to the docstrings of `save()` and `load()`, making it clear that the user should not expect robust behaviour of the load/save functionality at least during this development phase?

dcmckayibm commented:
@diego-plan9 see the updated commit based on your comments. I also ran things through `pylint`. 

The main change is that I pushed some of the file specific methods from the Results class to a new file in tools/file_io.py. This should also solve the issue you had with the Results initializer. 

I left in the datestr function (now in file_io.py) because, while I agree that it's not completely in scope to give this type of functionality to the user if we have a program making scientifically relevant data there should be a way to timestamp the filename. The function is easy and saves the user from writing it out and now it's out of the way in file_io.py.

I would make the same argument for your comment on `Results.save()` regarding stripping '.json' from the filename. We have to strip the .json anyway to append a number (in the case of duplicate files), so adding a check in there does not add extra code.

atilag commented:
Hi @dcmckayibm,
First of all thanks for the code!, @diego-plan9 did a great review so I don't really have much to say. There's still one thing that I think should be addressed for keeping the consistency of the SDK, and it's about moving the Save() functionality to `qiskit.tools`. There are two main reasons for justifying this change:
* `Result` is pretty simple class that it's meant to be a wrapper around the output of a circuit. It's part of what we consider the "core" of the SDK, so if we don't to break the [Single Responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle), this feature should be implemented somewhere else.
* Adding this functionality here, will "force" the users to save the results in this particular way, restricting other kind of use-cases for other users. For example, What if I want to save the results in XML, or YAML, or binary format?. This is a pretty nice feature that could be used by many users, but only if they want to, so moving it to our `qiskit.tools` package seems like the correct place to implement it.

Apologies for the inconveniences David! We are working on restructuring some things wishing that it will make contributions clearer! :)

dcmckayibm commented:
Ok @atilag I like that. Will move `save()`, however, I will need to add a method to `Results` to expose the two qobj dictionaries.

dcmckayibm commented:
Moved save to file_io.py and updated all the internal variables in `Results` to be 'internal' not 'private' (i.e. _ not __). Updated unit test accordingly and rebased.

diego-plan9 commented:
Thanks for the adjustments, @dcmckayibm, and also for taking care of moving `.__qobj` and `.__result` to private attributes after my not so detailed comment! It's looking ready to merge for me, with extra happiness due to fixing a number of style and docstring issues in the process - thank you!

ewinston commented on qiskit/tools/file_io.py:
```diff
@@ -0,0 +1,182 @@
+import qiskit
+import numpy
+
+def convert_qobj_to_json(in_item):
```
May be we should change name of this since this isn't actually converting a qobj to json but rather _result+_qobj, so we could just call it something like "_to_json()" like is done with python pandas for instance.
